File: LanguageWrappers/ruby/test_report_native.rb
#!/usr/bin/env ruby
# Unit tests for report_native.rb — the Ruby Fiddle wrapper for rdlnative.
#
# Requires a published rdlnative shared library.  Set RDLNATIVE_LIB to the
# path of rdlnative.so/.dylib/.dll before running, or the tests are skipped.
#
# Example (Linux):
#   dotnet publish RdlNative/... -p:PublishAot=true -o /tmp/rdlnative-pub
#   RDLNATIVE_LIB=/tmp/rdlnative-pub/rdlnative.so ruby test_report_native.rb

require 'minitest/autorun'
require 'tempfile'

LIB_PATH  = ENV.fetch('RDLNATIVE_LIB', '')
REPO_ROOT = File.expand_path('../../..', __FILE__)
RDL_PATH  = File.join(REPO_ROOT, 'Examples', 'SqliteExamples', 'SimpleTest1.rdl')
DB_PATH   = File.join(REPO_ROOT, 'Examples', 'northwindEF.db')
DB_CS     = "Data Source=#{DB_PATH}"

def library_available?
  LIB_PATH != '' && File.file?(LIB_PATH) &&
    File.file?(RDL_PATH) && File.file?(DB_PATH)
end

# Load wrapper and library once, lazily.
$fns = nil
def fns
  return $fns if $fns
  require_relative 'report_native'
  $fns = RdlLibrary.load(LIB_PATH)
end

# Helper: build a connected report object
def make_report
  require_relative 'report_native'
  rpt = ReportNative.new(fns, RDL_PATH)
  rpt.set_connection_string(DB_CS)
  rpt
end


class TestBasicRender < Minitest::Test
  def setup
    skip 'RDLNATIVE_LIB not set or library / sample files not found' unless library_available?
  end

  def test_pdf_memory
    data = make_report.export_to_memory('pdf')
    assert data.bytesize > 1000
    assert_equal '%PDF', data[0, 4].force_encoding('BINARY')
  end

  def test_html_memory
    data = make_report.export_to_memory('html')
    assert data.bytesize > 100
    assert_match(/<html/i, data.force_encoding('UTF-8'))
  end

  def test_csv_memory
    data = make_report.export_to_memory('csv')
    assert data.bytesize > 0
    assert_includes data.force_encoding('UTF-8'), 'Simple Test'
  end

  def test_xml_memory
    data = make_report.export_to_memory('xml')
    assert data.bytesize > 0
    assert_match(/<\?xml/i, data.force_encoding('UTF-8'))
  end

  def test_pdf_to_file
    Tempfile.create(['rdlnative_test', '.pdf']) do |f|
      path = f.path
      make_report.export('pdf', path)
      assert File.size(path) > 1000
    end
  end

  def test_multiple_renders_same_report
    rpt  = make_report
    pdf1 = rpt.export_to_memory('pdf')
    pdf2 = rpt.export_to_memory('pdf')
    assert_equal pdf1.bytesize, pdf2.bytesize
  end
end


class TestConnectionAndParameters < Minitest::Test
  def setup
    skip 'RDLNATIVE_LIB not set or library / sample files not found' unless library_available?
  end

  def test_set_connection_string
    require_relative 'report_native'
    rpt = ReportNative.new(fns, RDL_PATH)
    rpt.set_connection_string(DB_CS)
    data = rpt.export_to_memory('csv')
    assert_includes data.force_encoding('UTF-8'), 'Simple Test'
  end

  def test_set_parameter_does_not_crash
    require_relative 'report_native'
    rpt = ReportNative.new(fns, RDL_PATH)
    rpt.set_connection_string(DB_CS)
    rpt.set_parameter('SomeParam', 'SomeValue')
    data = rpt.export_to_memory('csv')
    assert data.bytesize > 0
  end
end


class TestErrorHandling < Minitest::Test
  def setup
    skip 'RDLNATIVE_LIB not set or library / sample files not found' unless library_available?
  end

  def test_invalid_rdl_path_raises
    require_relative 'report_native'
    rpt = ReportNative.new(fns, '/nonexistent/report.rdl')
    assert_raises(RuntimeError) { rpt.export_to_memory('pdf') }
  end

  def test_unknown_format_defaults_to_pdf
    data = make_report.export_to_memory('not_a_format')
    assert_equal '%PDF', data[0, 4].force_encoding('BINARY')
  end
end

File: LanguageWrappers/php/test_report_native.php
<?php
/**
 * Unit tests for report_native.php — the PHP FFI wrapper for rdlnative.
 *
 * Requires:
 *   - PHP FFI extension enabled (php.ini: extension=ffi, ffi.enable=true)
 *   - Published rdlnative shared library
 *   - RDLNATIVE_LIB environment variable set to the library path
 *   - LD_LIBRARY_PATH including the library directory (required before PHP starts
 *     because PHP's FFI loads with RTLD_LOCAL; .NET runtime libs must already
 *     be findable by the dynamic linker)
 *
 * Usage (Linux):
 *   dotnet publish RdlNative/... -p:PublishAot=true -o /tmp/rdlnative-pub
 *   export RDLNATIVE_LIB=/tmp/rdlnative-pub/rdlnative.so
 *   export LD_LIBRARY_PATH=/tmp/rdlnative-pub:$LD_LIBRARY_PATH
 *   php test_report_native.php
 *
 * Exit code 0 = all passed.  Each failure prints to stderr.
 */

declare(strict_types=1);

require_once __DIR__ . '/report_native.php';

use MajorsilenceReporting\RdlLibrary;
use MajorsilenceReporting\ReportNative;

// ─── Setup ────────────────────────────────────────────────────────────────────

$LIB_PATH  = (string)getenv('RDLNATIVE_LIB');
$REPO_ROOT = dirname(__DIR__, 2);
$RDL_PATH  = $REPO_ROOT . '/Examples/SqliteExamples/SimpleTest1.rdl';
$DB_PATH   = $REPO_ROOT . '/Examples/northwindEF.db';
$DB_CS     = "Data Source={$DB_PATH}";

function skip_if_unavailable(string $lib, string $rdl, string $db): void
{
    if ($lib === '') {
        fwrite(STDERR, "SKIP: RDLNATIVE_LIB not set.\n");
        exit(0);
    }
    if (!is_file($lib)) {
        fwrite(STDERR, "SKIP: RDLNATIVE_LIB={$lib} does not exist.\n");
        exit(0);
    }
    if (!is_file($rdl)) {
        fwrite(STDERR, "SKIP: Sample RDL not found at {$rdl}\n");
        exit(0);
    }
    if (!is_file($db)) {
        fwrite(STDERR, "SKIP: Sample DB not found at {$db}\n");
        exit(0);
    }
    if (!extension_loaded('ffi')) {
        fwrite(STDERR, "SKIP: PHP FFI extension not loaded.\n");
        exit(0);
    }
}

skip_if_unavailable($LIB_PATH, $RDL_PATH, $DB_PATH);

// ─── Mini test harness ────────────────────────────────────────────────────────

$passed = 0;
$failed = 0;

function run_test(string $name, callable $fn): void
{
    global $passed, $failed;
    try {
        $fn();
        echo "PASS  {$name}\n";
        $passed++;
    } catch (Throwable $e) {
        echo "FAIL  {$name}: {$e->getMessage()}\n";
        $failed++;
    }
}

function assert_true(mixed $value, string $msg = ''): void
{
    if (!$value) {
        throw new \AssertionError($msg ?: "Expected true, got false");
    }
}

function assert_greater(int $a, int $b, string $msg = ''): void
{
    if ($a <= $b) {
        throw new \AssertionError($msg ?: "Expected {$a} > {$b}");
    }
}

function assert_contains(string $needle, string $haystack, string $msg = ''): void
{
    if (strpos($haystack, $needle) === false) {
        throw new \AssertionError($msg ?: "Expected to find {$needle}");
    }
}

// ─── Shared library handle (loaded once) ─────────────────────────────────────

$ffi = null;
function get_ffi(): \FFI
{
    global $ffi, $LIB_PATH;
    if ($ffi === null) {
        $ffi = RdlLibrary::load($LIB_PATH);
    }
    return $ffi;
}

function make_report(): ReportNative
{
    global $RDL_PATH, $DB_CS;
    $rpt = new ReportNative(get_ffi(), $RDL_PATH);
    $rpt->set_connection_string($DB_CS);
    return $rpt;
}

// ─── Tests: Basic render ──────────────────────────────────────────────────────

run_test('test_pdf_memory', function () {
    $data = make_report()->export_to_memory('pdf');
    assert_greater(strlen($data), 1000);
    assert_true(substr($data, 0, 4) === '%PDF', 'Expected PDF magic bytes');
});

run_test('test_html_memory', function () {
    $data = make_report()->export_to_memory('html');
    assert_greater(strlen($data), 100);
    assert_contains('<html', strtolower($data));
});

run_test('test_csv_memory', function () {
    $data = make_report()->export_to_memory('csv');
    assert_greater(strlen($data), 0);
    assert_contains('Simple Test', $data);
});

run_test('test_xml_memory', function () {
    $data = make_report()->export_to_memory('xml');
    assert_greater(strlen($data), 0);
    assert_contains('<?xml', $data);
});

run_test('test_pdf_to_file', function () {
    $path = tempnam(sys_get_temp_dir(), 'rdlnative_') . '.pdf';
    try {
        make_report()->export('pdf', $path);
        assert_greater((int)filesize($path), 1000);
    } finally {
        if (file_exists($path)) unlink($path);
    }
});

run_test('test_multiple_renders_same_report', function () {
    $rpt  = make_report();
    $pdf1 = $rpt->export_to_memory('pdf');
    $pdf2 = $rpt->export_to_memory('pdf');
    if (strlen($pdf1) !== strlen($pdf2)) {
        throw new \AssertionError('Repeated render produced different sizes');
    }
});

// ─── Tests: Connection string and parameters ──────────────────────────────────

run_test('test_set_connection_string', function () {
    global $RDL_PATH, $DB_CS;
    $rpt = new ReportNative(get_ffi(), $RDL_PATH);
    $rpt->set_connection_string($DB_CS);
    $data = $rpt->export_to_memory('csv');
    assert_contains('Simple Test', $data);
});

run_test('test_set_parameter_does_not_crash', function () {
    $rpt = make_report();
    $rpt->set_parameter('SomeParam', 'SomeValue');
    $data = $rpt->export_to_memory('csv');
    assert_greater(strlen($data), 0);
});

// ─── Tests: Error handling ────────────────────────────────────────────────────

run_test('test_invalid_rdl_path_raises', function () {
    $rpt = new ReportNative(get_ffi(), '/nonexistent/report.rdl');
    $raised = false;
    try {
        $rpt->export_to_memory('pdf');
    } catch (\RuntimeException $e) {
        $raised = true;
        assert_greater(strlen($e->getMessage()), 0);
    }
    assert_true($raised, 'Expected RuntimeException for missing RDL');
});

run_test('test_unknown_format_defaults_to_pdf', function () {
    $data = make_report()->export_to_memory('not_a_format');
    assert_true(substr($data, 0, 4) === '%PDF', 'Unknown format should default to PDF');
});

// ─── Summary ──────────────────────────────────────────────────────────────────

echo "\n{$passed} passed, {$failed} failed.\n";
exit($failed > 0 ? 1 : 0);
